pennylane/gradients/param_shift_hessian.py

```diff
@@ -0,0 +1,320 @@
+        #         (h_dim, h_dim, out_dim) -> (out_dim, h_dim, h_dim)
+        hessian = qml.math.reshape(qml.math.stack(hessian), (h_dim, h_dim) + out_dim)
+        dim_indices = list(range(len(out_dim) + 2))
+        hessian = qml.math.transpose(hessian, axes=dim_indices[2:] + [0, 1])
+
+        return qml.math.squeeze(hessian)
+
+    return gradient_tapes, processing_fn
+
+
+@hessian_transform
+def param_shift_hessian(tape, f0=None):
+    r"""Transform a QNode to compute the parameter-shift Hessian with respect to its trainable
+    parameters.
+
+    Args:
+        tape (pennylane.QNode or .QuantumTape): quantum tape or QNode to differentiate
+        f0 (tensor_like[float] or None): Output of the evaluated input tape. If provided,
+            and the gradient recipe contains an unshifted term, this value is used,
+            saving a quantum evaluation.
+
+    Returns:
+        tensor_like or tuple[list[QuantumTape], function]:
+
+        - If the input is a QNode, a tensor representing the output of the hybrid Jacobian matrix
+          of size ``(QNode output dimensions, QNode input dimensions, QNode input dimensions)``
+          is returned. When the keyword ``hybrid=False`` is specified, the purely quantum Hessian
+          matrix is returned instead, with the dimesions
+          ``(QNode output dimensions, number of gate arguments, number of gate arguments)``.
+          The difference between the two accounts for the mapping of QNode arguments
+          to the actual gate arguments, which can include classical computations.
+
+        - If the input is a tape, a tuple containing a list of generated tapes, in addition
+          to a post-processing function to be applied to the evaluated tapes.
+    """
+
+    # perform gradient method validation
+    if any(m.return_type is qml.operation.State for m in tape.measurements):
+        raise ValueError(
+            "Computing the gradient of circuits that return the state is not supported."
+        )
+
+    # The parameter-shift Hessian implementation currently doesn't support variance measurements.
+    if any(m.return_type is qml.operation.Variance for m in tape.measurements):
+        raise ValueError(
+            "Computing the gradient of circuits that return variances is currently not supported."
+        )
+
+    if not tape.trainable_params:
+        return [], lambda _: []
+
+    # The parameter-shift Hessian implementation currently only supports
+    # the two-term parameter-shift rule. Raise an error for unsupported operations.
+    supported_ops = (
+        "RX",
+        "RY",
+        "RZ",
+        "Rot",
+        "PhaseShift",
+        "ControlledPhaseShift",
+        "MultiRZ",
+        "PauliRot",
+        "U1",
+        "U2",
+        "U3",
+        "SingleExcitationMinus",
+        "SingleExcitationPlus",
+        "DoubleExcitationMinus",
+        "DoubleExcitationPlus",
```

Comment:
Wait these ones have a two term rule? In my head, I kept thinking they had 4 term rules!

Reply:
Oh I actually didn't check these, I copied this list from here 😅
https://github.com/PennyLaneAI/pennylane/blob/c5d275eb825c1fe53844d26fe6275e198fc5a006/pennylane/tape/jacobian_tape.py#L764

Reply:
Nice! I suppose we could force an expansion of the tape if there are any operations with trainable parameters not in this list. But I wouldn't worry too much about this for now, since we have tools for a gen param shift coming up.